### PR TITLE
wrap CreateImage call in a retry to account for eventual consistencey…

### LINKED ISSUE #9562: Packer cannot wait for instance to be stopped
Hi:
 I encountered a problem. when i create a ami on aws,first packer will stop the instance and then copy the instance to a ami.but when packer stop the instance ,the instance will cost more than  five mins to stop. then i see that packer show error that "Error creating AMI: InvalidParameterValue: Invalid value 'i-05348a0d76476e460' for instanceId. Instance is not in state 'running' or 'stopped'
==> amazon-ebs: 	status code: 400, request id: *****".
so if there is a parameter to change the timeout value when packer wait for instance to be stop

below is the part of log:

2020/07/10 13:19:57 [INFO] (telemetry) ending shell
==> amazon-ebs: Stopping the source instance...
    amazon-ebs: Stopping instance
==> amazon-ebs: Waiting for the instance to stop...
==> amazon-ebs: Creating AMI test from instance i-05348a0d76476e460
==> amazon-ebs: 	status code: 400, request id: e1f57d04-c22b-4d2f-b50a-409fea3b1194
==> amazon-ebs: Error creating AMI: InvalidParameterValue: Invalid value 'i-05348a0d76476e460' for instanceId. Instance is not in state 'running' or 'stopped'
==> amazon-ebs: 	status code: 400, request id: e1f57d04-c22b-4d2f-b50a-409fea3b1194
==> amazon-ebs: Provisioning step had errors: Running the cleanup provisioner, if present...
==> amazon-ebs: Terminating the source AWS instance...
==> amazon-ebs: Cleaning up any extra volumes...
==> amazon-ebs: Destroying volume (vol-08a79b5bd61897fd6)...
==> amazon-ebs: Deleting temporary security group...
2020/07/10 13:26:35 [INFO] (telemetry) ending amazon-ebs
	status code: 400, request id: e1f57d04-c22b-4d2f-b50a-409fea3b1194
2020/07/10 13:26:35 machine readable: error-count []string{"1"}
==> Some builds didn't complete successfully and had errors:
2020/07/10 13:26:35 machine readable: amazon-ebs,error []string{"Error creating AMI: InvalidParameterValue: Invalid value 'i-05348a0d76476e460' for instanceId. Instance is not in state 'running' or 'stopped'\n\tstatus code: 400, request id: e1f57d04-c22b-4d2f-b50a-409fea3b1194"}
	status code: 400, request id: e1f57d04-c22b-4d2f-b50a-409fea3b1194
==> Builds finished but no artifacts were created.
2020/07/10 13:26:35 [INFO] (telemetry) Finalizing.
Build 'amazon-ebs' errored: Error creating AMI: InvalidParameterValue: Invalid value 'i-05348a0d76476e460' for instanceId. Instance is not in state 'running' or 'stopped'
	status code: 400, request id: e1f57d04-c22b-4d2f-b50a-409fea3b1194

### PR DESCRIPTION
**Pull request by SwampDragons**
wrap CreateImage call in a retry to account for eventual consistency issues with the image state. 

Closes #9562

### CODE REVIEW

File: builder/amazon/ebs/step_create_ami.go [hunk 1]
```diff
@@ -48,7 +51,22 @@ func (s *stepCreateAMI) Run(ctx context.Context, state multistep.StateBag) multi
...
 
-	createResp, err := ec2conn.CreateImage(createOpts)
+	var createResp *ec2.CreateImageOutput
+	var err error
+
+	err = retry.Config{
+		Tries: 11,
+		ShouldRetry: func(err error) bool {
+			if err, ok := err.(awserr.Error); ok {
+				return err.Code() == "InvalidParameterValue"
```
Comment (azr) [inline builder/amazon/ebs/step_create_ami.go:61]: `InvalidParameterValue` sounds like a non-retriable error but reading the logs is more :

`InvalidParameterValue: Invalid value 'i-05348a0d76476e460' for instanceId. Instance is not in state 'running' or 'stopped'`

LGTM !

Comment (azr) [reply builder/amazon/ebs/step_create_ami.go:61]: Ah we could use:

```go
			if isAWSErr(err, "InvalidParameterValue", "Instance is not in state") {
				return true
			}
```

Comment (SwampDragons) [reply builder/amazon/ebs/step_create_ami.go:61]: Yeah, that's probably better. I was being lazy and didn't want to convert isAwsErr to an exported value for this, but it's the right call.

File: builder/amazon/ebs/step_create_ami.go [hunk 2]
```diff
@@ -48,7 +50,22 @@ func (s *stepCreateAMI) Run(ctx context.Context, state multistep.StateBag) multi
 		BlockDeviceMappings: config.AMIMappings.BuildEC2BlockDeviceMappings(),
 	}
 
-	createResp, err := ec2conn.CreateImage(createOpts)
+	var createResp *ec2.CreateImageOutput
+	var err error
+
+	err = retry.Config{
+		Tries: 11,
```
Comment (HTTP500) [inline builder/amazon/ebs/step_create_ami.go:57]: Tried the test build and the error is slightly different:

```
ami: Error creating AMI: retry count exhausted. Last err: InvalidParameterValue: Invalid value 'i-00a737b5f100ce79d' for instanceId. Instance is not in state 'running' or 'stopped'\n==> ami: 	status code: 400%!(PACKER_COMMA) request id: 2144ba32-b620-4042-bf19-e7e6d522ee80
1594849969,,ui,say,==> ami: Provisioning step had errors: Running the cleanup provisioner%!(PACKER_COMMA) if present...
```

I tried setting `AWS_MAX_ATTEMPTS = 100` and `AWS_POLL_DELAY_SECONDS = 30` but still getting the same error.

I wonder if this line will make tries number to 11 no matter what you set to those 2 env vars?

Comment (SwampDragons) [reply builder/amazon/ebs/step_create_ami.go:57]: yeah the retry here is not related to the env vars, which get used when declaring waiters. 

I'm surprised though. If this were eventual consistency related, I'd have expected this wait loop to be enough. Can you set the environment variable PACKER_LOG=1 to turn on versbose logging and share full logs on the original issue? I'd also be interested to know how long Packer is spending around these lines: 

```
==> amazon-ebs: Stopping the source instance...
amazon-ebs: Stopping instance
==> amazon-ebs: Waiting for the instance to stop...
==> amazon-ebs: Creating AMI test from instance i-0dfeb326a14cb2aa4```

Comment (HTTP500) [reply builder/amazon/ebs/step_create_ami.go:57]: 2020/07/16 00:03:56 1594857836,,ui,say,==> ami: Stopping the source instance...
1594857836,,ui,say,==> ami: Stopping the source instance...
1594857836,,ui,message,    ami: Stopping instance
2020/07/16 00:03:56 1594857836,,ui,message,    ami: Stopping instance
2020/07/16 00:03:56 1594857836,,ui,say,==> ami: Waiting for the instance to stop...
1594857836,,ui,say,==> ami: Waiting for the instance to stop...
1594858168,,ui,say,==> ami: Creating AMI xxx from instance i-xxx
2020/07/16 00:09:28 1594858168,,ui,say,==> ami: Creating AMI xxx from instance i-xxx
2020/07/16 00:09:28 packer_test plugin: Retryable error: InvalidParameterValue: Invalid value 'i-xxx' for instanceId. Instance is not in state 'running' or 'stopped'
2020/07/16 00:09:28 packer_test plugin: 	status code: 400, request id: x
2020/07/16 00:09:29 packer_test plugin: Retryable error: InvalidParameterValue: Invalid value 'i-xxx' for instanceId. Instance is not in state 'running' or 'stopped'
2020/07/16 00:09:29 packer_test plugin: 	status code: 400, request id: x
2020/07/16 00:09:29 packer_test plugin: Retryable error: InvalidParameterValue: Invalid value 'i-xxx' for instanceId. Instance is not in state 'running' or 'stopped'
2020/07/16 00:09:29 packer_test plugin: 	status code: 400, request id: x
2020/07/16 00:09:30 packer_test plugin: Retryable error: InvalidParameterValue: Invalid value 'i-xxx' for instanceId. Instance is not in state 'running' or 'stopped'
2020/07/16 00:09:30 packer_test plugin: 	status code: 400, request id: x
2020/07/16 00:09:32 packer_test plugin: Retryable error: InvalidParameterValue: Invalid value 'i-xxx' for instanceId. Instance is not in state 'running' or 'stopped'
2020/07/16 00:09:32 packer_test plugin: 	status code: 400, request id: x
2020/07/16 00:09:35 packer_test plugin: Retryable error: InvalidParameterValue: Invalid value 'i-xxx' for instanceId. Instance is not in state 'running' or 'stopped'
2020/07/16 00:09:35 packer_test plugin: 	status code: 400, request id: x
2020/07/16 00:09:42 packer_test plugin: Retryable error: InvalidParameterValue: Invalid value 'i-xxx' for instanceId. Instance is not in state 'running' or 'stopped'
2020/07/16 00:09:42 packer_test plugin: 	status code: 400, request id: x
2020/07/16 00:09:55 packer_test plugin: Retryable error: InvalidParameterValue: Invalid value 'i-xxx' for instanceId. Instance is not in state 'running' or 'stopped'
2020/07/16 00:09:55 packer_test plugin: 	status code: 400, request id: x
2020/07/16 00:10:21 packer_test plugin: Retryable error: InvalidParameterValue: Invalid value 'i-xxx' for instanceId. Instance is not in state 'running' or 'stopped'
2020/07/16 00:10:21 packer_test plugin: 	status code: 400, request id: x
2020/07/16 00:10:51 packer_test plugin: Retryable error: InvalidParameterValue: Invalid value 'i-xxx' for instanceId. Instance is not in state 'running' or 'stopped'
2020/07/16 00:10:51 packer_test plugin: 	status code: 400, request id: x
2020/07/16 00:11:21 packer_test plugin: Retryable error: InvalidParameterValue: Invalid value 'i-xxx' for instanceId. Instance is not in state 'running' or 'stopped'
2020/07/16 00:11:21 packer_test plugin: 	status code: 400, request id: x
2020/07/16 00:11:51 1594858311,,ui,error,==> ami: Error creating AMI: retry count exhausted. Last err: InvalidParameterValue: Invalid value 'i-xxx' for instanceId. Instance is not in state 'running' or 'stopped'\n==> ami: 	status code: 400%!(PACKER_COMMA) request id: x
1594858311,,ui,error,==> ami: Error creating AMI: retry count exhausted. Last err: InvalidParameterValue: Invalid value 'i-xxx' for instanceId. Instance is not in state 'running' or 'stopped'\n==> ami: 	status code: 400%!(PACKER_COMMA) request id: x
1594858311,,ui,say,==> ami: Provisioning step had errors: Running the cleanup provisioner%!(PACKER_COMMA) if present...

Comment (SwampDragons) [reply builder/amazon/ebs/step_create_ami.go:57]: This is so weird.

Comment (HTTP500) [reply builder/amazon/ebs/step_create_ami.go:57]: Those EC2 metal boxes do take long time to power on/off. IMO we probably need longer waiting time and/or more retries

Comment (HTTP500) [reply builder/amazon/ebs/step_create_ami.go:57]: I believe we need longer waiting time / retries for this

I launched the packer build with the `-debug` flag so I could single step it. I waited 15 min after `StepStopEBSBackedInstance` and the build works

```
    ami: PLAY RECAP *********************************************************************
    ami: default                    : ok=12   changed=11   unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
    ami:
2020/07/20 20:32:49 packer-provisioner-ansible plugin: shutting down the SSH proxy
2020/07/20 20:32:49 [INFO] (telemetry) ending ansible
==> ami: Pausing after run of step 'StepProvision'. Press enter to continue. 
==> ami: Trying to remove ephemeral keys from authorized_keys files
2020/07/20 20:36:43 packer-builder-amazon-ebs plugin: [DEBUG] Opening new ssh session
2020/07/20 20:36:43 packer-builder-amazon-ebs plugin: [DEBUG] starting remote command: sed -i.bak '/ packer_5f15e763-7cae-3838-65a1-8df87abe8a67$/d' ~/.ssh/authorized_keys; rm ~/.ssh/authorized_keys.bak
2020/07/20 20:36:43 packer-builder-amazon-ebs plugin: [DEBUG] Opening new ssh session
2020/07/20 20:36:43 packer-builder-amazon-ebs plugin: [DEBUG] starting remote command: sudo sed -i.bak '/ packer_5f15e763-7cae-3838-65a1-8df87abe8a67$/d' /root/.ssh/authorized_keys; sudo rm /root/.ssh/authorized_keys.bak
==> ami: Pausing after run of step 'StepCleanupTempKeys'. Press enter to continue. 
==> ami: Stopping the source instance...
    ami: Stopping instance
==> ami: Waiting for the instance to stop...
==> ami: Pausing after run of step 'StepStopEBSBackedInstance'. Press enter to continue. 

(I waited 15min here)

==> ami: Pausing after run of step 'StepModifyEBSBackedInstance'. Press enter to continue. 
==> ami: Pausing after run of step 'StepDeregisterAMI'. Press enter to continue. 
==> ami: Creating AMI xxx from instance i-xxx
    ami: AMI: ami-xxx
==> ami: Waiting for AMI to become ready...
==> ami: Pausing after run of step 'stepCreateAMI'. Press enter to continue. 
==> ami: Pausing after run of step 'StepAMIRegionCopy'. Press enter to continue. 
==> ami: Pausing after run of step 'StepModifyAMIAttributes'. Press enter to continue. 
==> ami: Adding tags to AMI (ami-xxx)...
==> ami: Tagging snapshot: snap-xxx
```

Comment (azr) [reply builder/amazon/ebs/step_create_ami.go:57]: Ah, nice, setting tries to `0` will not limit on the number of retires but will leave when ctx is cancelled. May be creating a context with a timeout of 15min will solve this:
```suggestion
		Tries: 0,
```

Comment (SwampDragons) [reply builder/amazon/ebs/step_create_ami.go:57]: The thing that's confusing me isn't that it takes a long time to stop -- it's that the waiter being called in the part of the code that says "Waiting for the instance to stop..." should be the one that waits a long time and has to do a bunch of retries, not this "creating image" step. We need to figure out why that waiter is passing when the instance clearly isn't stopped.

Comment (HTTP500) [reply builder/amazon/ebs/step_create_ami.go:57]: IIRC When I was doing single step, I saw `==> ami: Pausing after run of step 'StepStopEBSBackedInstance'. Press enter to continue. ` right after AWS console showing the instance is stopped (I am refreshing page every few sec). So IMO the waiter did nothing wrong, it is just some special behaviour of AWS's metal instances

Comment (SwampDragons) [reply builder/amazon/ebs/step_create_ami.go:57]: So maybe for some reason the eventual consistency is more... "eventual" with the metal instances. Okay. I'll tweak this code again to make the retries last longer.

**Comment by SwampDragons** [comment]
I still can't reproduce this on my end, but I've pushed up a change to set the timeout on this retry loop to 15 minutes; new builds are here: https://circleci.com/gh/hashicorp/packer/68414#artifacts/containers/0

File: builder/amazon/ebs/step_create_ami.go (hunk 2 above)
Comment (HTTP500) [reply builder/amazon/ebs/step_create_ami.go]:
```
2020/07/22 00:55:31 1595379331,,ui,say,==> ami: Waiting for the instance to stop...
1595379331,,ui,say,==> ami: Waiting for the instance to stop...
1595379664,,ui,say,==> ami: Creating AMI xxx from instance xxx
2020/07/22 01:01:04 1595379664,,ui,say,==> ami: Creating AMI xxx from instance xxx
1595379664,,ui,message,    ami: AMI: ami-xxx
2020/07/22 01:01:04 1595379664,,ui,message,    ami: AMI: ami-xxx
2020/07/22 01:01:04 1595379664,,ui,say,==> ami: Waiting for AMI to become ready...
```
It works!
Thanks @SwampDragons @azr !

**Comment by HTTP500** [comment]
```
2020/07/22 00:55:31 1595379331,,ui,say,==> ami: Waiting for the instance to stop...
1595379331,,ui,say,==> ami: Waiting for the instance to stop...
1595379664,,ui,say,==> ami: Creating AMI xxx from instance xxx
2020/07/22 01:01:04 1595379664,,ui,say,==> ami: Creating AMI xxx from instance xxx
1595379664,,ui,message,    ami: AMI: ami-xxx
2020/07/22 01:01:04 1595379664,,ui,message,    ami: AMI: ami-xxx
2020/07/22 01:01:04 1595379664,,ui,say,==> ami: Waiting for AMI to become ready...
```
It works!
Thanks @SwampDragons @azr !

**Comment by SwampDragons** [comment]
Glad to hear it. Thanks for testing for us

**Pull request merged**